### PR TITLE
Quasi-Newton Models

### PR DESCRIPTION
**Pull request by dpo**
A simple initial design for quasi-Newton models. Example
```julia
julia> using NLPModels
julia> model = SimpleNLPModel(x -> dot(x, x), zeros(2))
julia> qnmodel = LBFGSModel(model)  # or LSR1Model; optional memory argument
```
Differences with other types of models:
* `hess_coord()` is not implemented (the Hessian is typically dense)
* `hess()` ~~returns a `LinearOperator`~~ is not implemented
* `hprod()` applies the linear operator

I'll add tests when we agree on the design.

@abelsiqueira

### CODE REVIEW

**Comment by abelsiqueira** [comment]
Very good. I think the update with a separate function is best. Something like `update!(nlp)`. It should be implemented by default for an AbstractNLPModel so that calling it for a non-qn model won't raise anything.
Another thing, I think `hess` should also be undefined, because it should returns a triangular matrix.
Also, `reset!` should reset the operator too.

**Comment by dpo** [comment]
Ok, thanks. A few points to sort out:

* I suggest that all `AbstractNLPModel`s have a new method, say `hess_op()` that returns the Hessian as a linear operator (based on `hprod()`). `QuasiNewtonModel`s will implement `hess_op()` and not `hess()`. Does that make sense?
* does a `QuasiNewtonModel` really need its own `counters` field? I'm thinking not, but then we need an API to access the counters across model types.
* ok for an `update!()` method.

**Comment by abelsiqueira** [comment]
> I suggest that all AbstractNLPModels have a new method, say hess_op() that returns the Hessian as a linear operator (based on hprod()). QuasiNewtonModels will implement hess_op() and not hess(). Does that make sense?

Yes, and strongly agree with `hess_op`.
> does a QuasiNewtonModel really need its own counters field? I'm thinking not, but then we need an API to access the counters across model types.

Does `hprod` on a QNM increases the `hprod` counter of the inner model?

**Comment by dpo** [comment]
> Does hprod on a QNM increases the hprod counter of the inner model?

No because an actual hprod was never computed. But perhaps `LinearOperators` should also contain a counter?!

**Comment by abelsiqueira** [comment]
Don't know if a counter in LO would be beneficial for everyone. But a counter `Bprod` inside the QuasiNewtonModel might. Either way, that's an additional counter to worry about.
Otherwise, I might say the `counters(nlp)` returning a correct counter should be enough. It still might for most cases, I think.

Possibly, someone actually interested in `Bprod` will be careful enough to only look for Bprods inside QNMs?

**Comment by dpo** [comment]
> Possibly, someone actually interested in Bprod will be careful enough to only look for Bprods inside QNMs?

Possibly, but they would only be counting a handful of dot products per hprod.

**Comment by abelsiqueira** [comment]
So no `Bprod`? In that case, I mantain `counters(nlp)`.

**Comment by dpo** [comment]
`SlackModel`s don't have a `counter` either because the underlying nlp is already counting. I feel we should have an API of the form
```julia
neval_obj(nlp :: AbstractNLPModel) = nlp.counters.neval_obj
# etc.
```
so that we can define
```julia
neval_obj(nlp :: SlackModel) = neval_obj(nlp.model)
neval_obj(nlp :: QuasiNewtonModel) = neval_obj(nlp.model)
# etc.
```
What do you think?

**Comment by abelsiqueira** [comment]
That's even better.

**Comment by dpo** [comment]
At this point, I'm too tired to figure out where this error is coming from...
```
Checking LBFGS formulation of hs10     	ERROR: LoadError: LoadError: push! not implemented
```

File: src/qn_model.jl [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+for counter in fieldnames(Counters)
+  @eval begin
+    $counter(nlp :: QuasiNewtonModel) = $counter(nlp.model)
+    export $counter
+  end
+end
+
+function reset!(nlp :: QuasiNewtonModel)
+  reset!(nlp.model.counters)
+  return nlp
```
Comment (abelsiqueira) [inline src/qn_model.jl]: Can we reset the Operator here?

Comment (dpo) [inline src/qn_model.jl]: Great question! That's not currently possible but it's necessary!

Comment (abelsiqueira) [inline src/qn_model.jl]: Done in JuliaSmoothOptimizers/LinearOperators.jl/pull/16.
`LinearOperators.reset!(nlp.op)` should be enough.

File: test/test_qn_model.jl [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  v = rand(qnmodel.meta.nvar)
+  u = rand(qnmodel.meta.ncon)
+
+  @assert all(jprod(model, x, v) == jprod(qnmodel, x, v))
+  @assert neval_jprod(model) == 2
+
+  @assert all(jtprod(model, x, u) == jtprod(qnmodel, x, u))
+  @assert neval_jtprod(model) == 2
+
+  H = hess_op(model, x)
```
Comment (abelsiqueira) [inline test/test_qn_model.jl]: This should be `qnmodel`

File: test/test_qn_model.jl [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  @assert typeof(H) <: LinearOperators.AbstractLinearOperator
+  @assert size(H) == (model.meta.nvar, model.meta.nvar)
+  @assert all(H * v == hprod(model, x, v))
```
Comment (abelsiqueira) [inline test/test_qn_model.jl]: This should also be `qnmodel` and `==` verifies all components. `all` is not necessary.

**Comment by dpo** [comment]
Updated.

File: src/qn_model.jl [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+for counter in fieldnames(Counters)
+  @eval begin
+    $counter(nlp :: QuasiNewtonModel) = $counter(nlp.model)
+    export $counter
+  end
+end
+
+function reset!(nlp :: QuasiNewtonModel)
+  reset!(nlp.model.counters)
+  reset!(nlp.op)
```
Comment (abelsiqueira) [inline src/qn_model.jl]: `LinearOperators.reset!`

Comment (dpo) [inline src/qn_model.jl]: This is mysterious to me. There's a `using LinearOperators` at the top of `NLPModels.jl` (which `include`s `qn_model.jl`). Why doesn't multiple dispatch do its job here?

Comment (dpo) [inline src/qn_model.jl]: We have to explicitly `import` the methods we override.

Comment (abelsiqueira) [inline src/qn_model.jl]: Now it's failing because `LinearOperators.reset!` is in the latest develop.

Comment (dpo) [inline src/qn_model.jl]: Right so we need a new release of LinearOperators. Do you see anything else you'd like to get in there very soon?

Comment (abelsiqueira) [inline src/qn_model.jl]: Nothing in mind.

**Comment by dpo** [comment]
Hm. I wonder why Travis and Appveyor aren't kicking in.

**Comment by dpo** [comment]
How can the same build options behave differently on different repos?!

**Comment by dpo** [comment]
Seems like a TravisCI bug.

**Comment by dpo** [comment]
I think this is happy now.

**Comment by abelsiqueira** [comment]
Thank you

**Pull request merged**